src/ntpclient.cpp

```diff
@@ -0,0 +1,195 @@
+            {
+
+                LogPrint("ntp", "[NTP] Could not read socket from NTP server %s (Read timeout)\n", sHostName);
+
+            }
+            else
+            {
+
+                socket.receive_from(boost::asio::buffer(recvBuf), sender_endpoint);
+                timeRecv = ntohl((time_t)recvBuf[4]);
+                timeRecv-= 2208988800U;  // Substract 01/01/1970 == 2208988800U
+
+                LogPrint("ntp", "[NTP] Received timestamp: %ll \n", (uint64_t)timeRecv);
+
+            }
+
+        }
+        catch (std::exception& e)
+        {
+
+             LogPrintf("[NTP] Could not read clock from NTP server %s (%s)\n", sHostName, e.what());
+
+        }
+
+    }
+    catch (std::exception& e)
+    {
+
+        LogPrintf("[NTP] Could not open socket to NTP server %s (%s)\n", sHostName, e.what());
+
+    }
+
+    return (int64_t)timeRecv;
+}
+
+bool NtpClockSync() {
+    LogPrintf("[NTP] Starting clock sync...\n");
+
+    std::vector<std::string> vNtpServers;
+    std::vector<int64_t> vResults;
+
+    if (mapMultiArgs["-ntpservers"].size() > 0)
+    {
+        vNtpServers = mapMultiArgs["-ntpservers"];
+    }
+    else
+    {
+        vNtpServers = vDefaultNtpServers;
+    }
+
+    string sReport = "";
+    string sPrevServer = "";
+    int64_t nPrevMeasure = -1;
+
+    random_shuffle(vNtpServers.begin(), vNtpServers.end(), GetRandInt);
+
+    for(unsigned int i = 0; i < vNtpServers.size(); i++)
+    {
+        string s = vNtpServers[i];
+        CNtpClient ntpClient(s);
+        int64_t nTimestamp = ntpClient.getTimestamp();
+        if(nTimestamp > -1)
+        {
+            int64_t nClockDrift = GetTimeNow() - nTimestamp;
+
+            // We push if its the first entry
+            if(vResults.size() == 0)
+            {
+                vResults.push_back(nClockDrift);
+
+                string sign = "";
+                if(nClockDrift > 0)
+                    sign = "+";
+                else if (nClockDrift < 0)
+                    sign = "-";
+
+                sReport += s + "[" + sign + to_string(nClockDrift) + "sec.] ";
+            }
+            // or if we have two cached values, ensuring size of the vector is odd
+            else if (nPrevMeasure != -1)
+            {
+                vResults.push_back(nClockDrift);
+
+                string sign = "";
+                if(nClockDrift > 0)
+                    sign = "+";
+                else if (nClockDrift < 0)
+                    sign = "-";
+
+                sReport += s + "[" + sign + to_string(nClockDrift) + "sec.] ";
+
+                vResults.push_back(nPrevMeasure);
+
+                sign = "";
+                if(nPrevMeasure > 0)
+                    sign = "+";
+                else if (nPrevMeasure < 0)
+                    sign = "-";
+
+                sReport += sPrevServer + "[" + sign + to_string(nPrevMeasure) + "sec.] ";
+
+                nPrevMeasure = -1;
+                sPrevServer = "";
+            }
+            else
```

Comment:
Would this logic be more readable if it were to work on a modulus rather than setting and unsetting the nPrevMeasure variable to -1 ? It took me a while to understand what was going in here and what happened during each cycle of the for loop.

```
if (i == 0)
    push_back(first_measure)
else if (i % 2 == 1)
    push_back(previous_measure)
    push_back(current_measure)
else 
    previous_measure = current_measure
```

Reply:
Done -> 5840155a7f6dfc8f6f84e34424f16e501fd62c67